### PR TITLE
New Exception ctor that takes the name of a type and object.

### PR DESCRIPTION
**Pull request by chrisdembia**
It's very common that we want to throw an exception in which the message contains the name of the class and the name of an instance of that class. It's somewhat tedious to create error messages that include these names, and we are often inconsistent about how we format such messages.

This exception constructor allows us to include such names in exception messages more easily and uniformly.

@klshrinidhi

### CODE REVIEW

File: OpenSim/Common/Exception.cpp [hunk 1]
```diff
@@ -70,8 +70,21 @@ exception()
...
-    addMessage(file + ":" + std::to_string(line) + "\n" + "In function '" + 
-               func + "'");
+    addMessage("\tIn file " + file + ":" + std::to_string(line) + "\n" +
+               "\tIn function '" + func + "'");
+}
+
+Exception::Exception(const std::string& file,
+              size_t line,
+              const std::string& func,
+              const std::string& className,
```
Comment (aseth1) [inline OpenSim/Common/Exception.cpp]: What do you think of passing a const ref to the object itself? And strip the className and objectName out for us?

Comment (chrisdembia) [inline OpenSim/Common/Exception.cpp]: Okay yeah. That will work for all Objects.

**Comment by klshrinidhi** [comment]
When added the previous constructor, I wanted to add the class name too. I didn't add it at the time because there was no standard way to automatically infer class name at the throw site using macros like I have [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Exception.h#L62). It had to be inferred by demangling result of `typeid`. There is a library to do this. I wouldn't like it if throw sites had to specify class name all the time.

**Comment by chrisdembia** [comment]
`SimTK::NiceTypeName<T>::namestr()` returns the demangled name.

> There is a library to do this. I wouldn't like it if throw sites had to specify class name all the time.

I agree.

What about the name associated with an Object/Component (not its type name, its instance name)?

**Comment by chrisdembia** [comment]
Should we hold off on this PR then?

**Comment by chrisdembia** [comment]
Merge conflict resolved.

**Comment by klshrinidhi** [comment]
I personally would hold off on this until compilers give us macros to do this. I think the *file-name, function name and line-number* are enough information until then. 
If you want to go ahead with this, I think there will be need for another macro `OPENSIM_THROW_<something>` that uses this constructor. Every throw site will have to explicitly disambiguate by calling the right macro.

**Comment by klshrinidhi** [comment]
:+1: 
I think `OPENSIM_THROW_<something>` macro to pick up this constructor is the only thing left here.

**Comment by klshrinidhi** [comment]
I have:
* Added another constructor similar to the one Chris added. This constructor is intended for derived classes. That makes a pair of these constructors just like there was a pair before that does not take object.
* I have added two macros for throw sites to use. The macros pick up the `this` pointer so throw sites do not have to specify explicitly. I mentioned this in my first comment on this PR

Take a look at the macros. I have not used them yet.

**Comment by chrisdembia** [comment]
Looks good to me! Thank you for doing this.

**Comment by klshrinidhi** [comment]
Thanks for convincing me.
All :white_check_mark: !!!
Merging now.

**Pull request merged**